content/posts/networkx/node-ordering-Ti-updating/index.md

```diff
@@ -0,0 +1,157 @@
+
+## $T_i$ and $\tilde{T_i}$
+
+According to the VF2++ paper notation:
+
+$$T_1=\(u\in V_1 \setminus m: \exists \tilde{u} \in m: (u,\tilde{u}\in E_1)\)$$
+
+where $V_1$ and $E_1$ contain all the nodes and edges of the first graph respectively, and $m$ is a dictionary, mapping
+every node of the first graph to a node of the second graph. Now if we interpet the above equation, we conclude that
+$T_1$ contains uncovered neighbors of covered nodes. In simple terms, it includes all the nodes that do not belong in
+the mapping $m$ yet, but are neighbors of nodes that are in the mapping. In addition,
+
+$$\tilde{T_1}=\(V_1 \setminus m \setminus T_1\)$$
+
+The following figure is meant to provide some visual explanation of what exactly $T_i$ is.
+
+![Illustration of $T_i$.](Ti.png)
+
+The blue nodes 1,2,3 are nodes from graph G1 and the green nodes A,B,C belong to the graph G2. The grey lines connecting
+those two indicate that in this current state, node 1 is mapped to node A, node 2 is mapped to node B etc. The yellow
+edges are just the neighbors of the covered (mapped) nodes. Here, $T_1$ contains the red nodes (4,5,6) which are
+neighbors of the covered nodes 1,2,3, and $T_2$ contains the grey ones (D,E,F). None of the nodes depicted would be
+included in $\tilde{T_1}$ or $\tilde{T_2}$. The latters would contain all the remaining nodes from the two graphs.
+
+Regarding the computation of these sets, it's not practical to use the brute force method and iterate over all nodes in
+every step of the algorithm to find the desired nodes and compute $T_i$ and $\tilde{T_i}$. We use the following
+observations to implement an incremental computation of $T_i$ and $\tilde{T_i}$ and make VF2++ more efficient.
+
+- $T_i$ is empty in the beggining, since there are no mapped nodes ($m=\varnothing$) and therefore no neighbors of
+  mapped nodes.
+- $\tilde{T_i}$ initially contains all the nodes from graph $G_i, i=1,2$ which can be realized directly from the
+  notation if we consider both $m$ and $T_1$ empty sets.
+- Every step of the algorithm either adds one node $u$ to the mapping or pops one from it.
+
+We can conclude that in every step, $T_i$ and $\tilde{T_i}$ can be incrementally updated. This method avoids a ton of
+redundant operations and results in significant performance improvement.
+
+![Performance comparison between brute force Ti computing and incremental updating.](acceleration.png)
+
+The above graph shows the difference in performance between using the exhaustive brute force and incrementally updating
+$T_i$ and $\tilde{T_i}$. The graph used to obtain these measurements was a regular
+[GNP Graph](https://en.wikipedia.org/wiki/Erd%C5%91s%E2%80%93R%C3%A9nyi_model) with a probability for an edge equal to
+$0.7$. It is directly observed that the increase in time when using the brute force method, follows an exponential rate,
+while in our method, it's not significantly affected by incrementing the number of nodes. Let's talk time complexity.
```

Comment:
I think this statement doesn't *quite* match what the performance plot shows. For instance, both methods *appear* to have an exponential increase in run-time vs. number of nodes, it's just that the incremental update method increases much less rapidly. Perhaps something like:

```suggestion
$0.7$. It can clearly be seen that execution time of the brute force
method increases much more rapidly with the number of nodes/edges than
the incremental update method, as expected.
```

Reply:
Sure, you are right!